Программа распознавания языка. На текстах википедии проверяем два метода: словарный метод и N-граммы.

### Шаг1. Скачать тренировочные тексты из Вики

In [1]:
import wikipedia

In [2]:
#функция скачивает из Википедии n текстов на языке lang
def get_texts_for_lang(lang, n=10):
    wikipedia.set_lang(lang)
    wiki_content = []
    pages = wikipedia.random(n)
    for page_name in pages:
        try:
            page = wikipedia.page(page_name)
        except wikipedia.exceptions.WikipediaException:
            print('Skipping page {}'.format(page_name))
            continue

        wiki_content.append('{}\n{}'.format(page.title, page.content.replace('==', '')))

    return wiki_content

In [3]:
# скачиваем по 100 статей для каждого языка
#kk - казахский, uk - украинский, be - белорусский, fr - французский

wiki_texts = {}

In [4]:
wiki_texts['kk'] = get_texts_for_lang('kk', 100)
print('kk', len(wiki_texts['kk']))

//anaconda/lib/python3.5/site-packages/bs4/__init__.py:181: UserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("lxml"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 184 of the file //anaconda/lib/python3.5/runpy.py. To get rid of this warning, change code that looks like this:

 BeautifulSoup([your markup])

to this:

 BeautifulSoup([your markup], "lxml")

  markup_type=markup_type))


Skipping page Мәдениет ауылдық округі
kk 99


In [5]:
wiki_texts['uk'] = get_texts_for_lang('uk', 100)
print('uk', len(wiki_texts['uk']))

//anaconda/lib/python3.5/site-packages/bs4/__init__.py:181: UserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("lxml"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 184 of the file //anaconda/lib/python3.5/runpy.py. To get rid of this warning, change code that looks like this:

 BeautifulSoup([your markup])

to this:

 BeautifulSoup([your markup], "lxml")

  markup_type=markup_type))


Skipping page Костерін
Skipping page Інтерв'ю (значення)
Skipping page Флореску
Skipping page Симонів
Skipping page Складан
Skipping page Торчин
Skipping page Євсєєв
uk 93


In [6]:
wiki_texts['be'] = get_texts_for_lang('be', 100)
print('be', len(wiki_texts['be']))

//anaconda/lib/python3.5/site-packages/bs4/__init__.py:181: UserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("lxml"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 184 of the file //anaconda/lib/python3.5/runpy.py. To get rid of this warning, change code that looks like this:

 BeautifulSoup([your markup])

to this:

 BeautifulSoup([your markup], "lxml")

  markup_type=markup_type))


Skipping page Грынявічы
Skipping page Канюхі
Skipping page Амасія
be 97


In [7]:
wiki_texts['fr'] = get_texts_for_lang('fr', 100)
print('fr', len(wiki_texts['fr']))

//anaconda/lib/python3.5/site-packages/bs4/__init__.py:181: UserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("lxml"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 184 of the file //anaconda/lib/python3.5/runpy.py. To get rid of this warning, change code that looks like this:

 BeautifulSoup([your markup])

to this:

 BeautifulSoup([your markup], "lxml")

  markup_type=markup_type))


Skipping page Delbet
Skipping page Croix de guerre
Skipping page Lahic
Skipping page Association Sportive de Corbeil-Essonnes Canoë-Kayak
Skipping page Liu Zhihua
Skipping page Conil
Skipping page Cottingham
fr 93


In [17]:
# распечатаем по 1 тексту
print(wiki_texts['kk'][0])
print(wiki_texts['uk'][0])
print(wiki_texts['be'][0])
print(wiki_texts['fr'][0])

Аврамово (Кырджали облысы)
Аврамово (болг. Аврамово) — Болгариядағы ауыл. Кырджали облысына қарасты Добрич қауымының құрамына кіреді.
2010 жылғы мәліметтер бойынша тұрғындарының саны 5 адамды құрайды.
Пошта индексі — 6797. Телефон коды — 07442. ЕКАТТЕ коды — 62.


 Картада 
bgmaps.com электронды картада орналасуы
emaps.bg электронды картада орналасуы
Google электронды картада орналасуы


 Сыртқы сілтемелер 
Тұрғын статистикасы  (болг.)
Іоанніс Ралліс
Іоа́нніс Ра́лліс (грец. Ιωάννης Ράλλης; 1878 — 26 жовтня 1946) — грецький політик, глава лялькового уряду часів німецької окупації Греції.


 Життєпис 
Належав до відомої політичної династії (його батько також був прем'єр-міністром). 1936 року об'єднався з колишнім диктатором Георгіосом Кондилісом, якого король Георг II усунув від влади за кілька днів після своєї реставрації. Був засуджений за зраду до довічного ув'язнення, помер у в'язниці.
Його син, Георгіос, 1947 року видав книгу «Іоанніс Ралліс мовить з могили», де було посмертно опубл

# Метод 1. Словарный метод

### Шаг 2. Составить частотные списки

In [18]:
import codecs
import collections
import sys

def tokenize(text): #токенизация по пробелам
    return text.split(' ')

In [221]:
freqs_kk = collections.defaultdict(lambda: 0)
corpus_kk = wiki_texts['kk']
for article in corpus_kk:
    for word in tokenize(article.replace('\n', '').lower()):
        freqs_kk[word] += 1

for word in sorted(freqs_kk, key=lambda w: freqs_kk[w], reverse=True)[:50]:
    print('{}\t{}'.format(freqs_kk[word], word))

In [222]:
freqs_uk = collections.defaultdict(lambda: 0)
corpus_uk = wiki_texts['uk']
for article in corpus_uk:
    for word in tokenize(article.replace('\n', '').lower()):
        freqs_uk[word] += 1

for word in sorted(freqs_uk, key=lambda w: freqs_uk[w], reverse=True)[:50]:
    print('{}\t{}'.format(freqs_uk[word], word))

In [223]:
freqs_be = collections.defaultdict(lambda: 0)
corpus_be = wiki_texts['be']
for article in corpus_be:
    for word in tokenize(article.replace('\n', '').lower()):
        freqs_be[word] += 1

for word in sorted(freqs_be, key=lambda w: freqs_be[w], reverse=True)[:50]:
    print('{}\t{}'.format(freqs_be[word], word))

In [224]:
freqs_fr = collections.defaultdict(lambda: 0)
corpus_fr = wiki_texts['fr']
for article in corpus_fr:
    for word in tokenize(article.replace('\n', '').lower()):
        freqs_fr[word] += 1

for word in sorted(freqs_fr, key=lambda w: freqs_fr[w], reverse=True)[:50]:
    print('{}\t{}'.format(freqs_fr[word], word))

Почистим словари от слов, которые встречаются больше, чем в 1 языке. Таким образом, у нас останутся только лингвоспецифические слова.

In [250]:
freqs_list = [freqs_kk, freqs_uk, freqs_be, freqs_fr]

In [251]:
for i in range(0, len(freqs_list)):
    words = list(freqs_list[i].keys())
    k = 0
    while k != len(words):
        word = words[k]
        for j in range(0, len(freqs_list)):
            if freqs_list[j] != freqs_list[i]:
                if word in freqs_list[j]:
                    del freqs_list[j][word]
                    flag = 'T'
        if flag == 'T':
            del freqs_list[i][word]
            flag = 'F'
        k += 1

Сделаем из них словари по 100 самых частотных слов:

In [253]:
def freqs_sort(d, n):
    freqs_sort = {}
    words_sort = sorted(d, key=lambda w: d[w], reverse=True)[:n]
    for i in words_sort:
        freqs_sort[i] = d[i]
    return freqs_sort

In [254]:
freqs_kk_sort = freqs_sort(freqs_kk, 100)
freqs_uk_sort = freqs_sort(freqs_uk, 100)
freqs_be_sort = freqs_sort(freqs_be, 100)
freqs_fr_sort = freqs_sort(freqs_fr, 100)

### Шаг 3. Определить язык

Делим текст на токены и смотрим, в каком частотном словаре больше слов из текста 

In [370]:
def lng_id_words(text, freqs, lngs): #freqs - все имеющиеся частотные словари
    count_freqs = [0]*len(freqs)
    text_tokenized = tokenize(text.replace('\n', '').lower())
    for i in text_tokenized:
        for j in range(0, len(freqs)):
            if i in freqs[j]:
                count_freqs[j] += 1
    return lngs[count_freqs.index(max(count_freqs))]

### Шаг 4. Скачать тестовые тексты из Википедии

In [45]:
wiki_texts_test = {}

In [46]:
wiki_texts_test['kk'] = get_texts_for_lang('kk', 100)
print('kk', len(wiki_texts_test['kk']))

//anaconda/lib/python3.5/site-packages/bs4/__init__.py:181: UserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("lxml"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 184 of the file //anaconda/lib/python3.5/runpy.py. To get rid of this warning, change code that looks like this:

 BeautifulSoup([your markup])

to this:

 BeautifulSoup([your markup], "lxml")

  markup_type=markup_type))


Skipping page Седан
kk 99


In [47]:
wiki_texts_test['uk'] = get_texts_for_lang('uk', 100)
print('uk', len(wiki_texts_test['uk']))

//anaconda/lib/python3.5/site-packages/bs4/__init__.py:181: UserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("lxml"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 184 of the file //anaconda/lib/python3.5/runpy.py. To get rid of this warning, change code that looks like this:

 BeautifulSoup([your markup])

to this:

 BeautifulSoup([your markup], "lxml")

  markup_type=markup_type))


Skipping page Фрессін
Skipping page Вовківська сільська рада
Skipping page Чиж (значення)
Skipping page Підлісне (Белебеївський район)
Skipping page Оболонська вулиця
Skipping page Микитенко Ольга Леонтіївна
Skipping page Ягорі
Skipping page Шеховцов Анатолій Федорович
uk 92


In [260]:
wiki_texts_test['be'] = get_texts_for_lang('be', 100)
print('be', len(wiki_texts_test['be']))

//anaconda/lib/python3.5/site-packages/bs4/__init__.py:181: UserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("lxml"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 184 of the file //anaconda/lib/python3.5/runpy.py. To get rid of this warning, change code that looks like this:

 BeautifulSoup([your markup])

to this:

 BeautifulSoup([your markup], "lxml")

  markup_type=markup_type))


Skipping page Гарбачы
Skipping page Лік, значэнні
Skipping page Ніва
Skipping page Пракопавічы
Skipping page Грэск
be 95


In [261]:
wiki_texts_test['fr'] = get_texts_for_lang('fr', 100)
print('fr', len(wiki_texts_test['fr']))

//anaconda/lib/python3.5/site-packages/bs4/__init__.py:181: UserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("lxml"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 184 of the file //anaconda/lib/python3.5/runpy.py. To get rid of this warning, change code that looks like this:

 BeautifulSoup([your markup])

to this:

 BeautifulSoup([your markup], "lxml")

  markup_type=markup_type))


Skipping page ADY
Skipping page Missisicabi
Skipping page Ligne 108
Skipping page Oberglatt
Skipping page Pas sur la bouche (homonymie)
Skipping page Luini (homonymie)
fr 94


### Шаг 5. Проверить метод на тестовой выборке

In [379]:
freqs_words_sort_list = [freqs_kk_sort, freqs_uk_sort, freqs_be_sort, freqs_fr_sort]
lngs = ['Kazakh', 'Ukranian', 'Belorussian', 'French']

In [380]:
corpus_kk_test = wiki_texts_test['kk']
kk = 0
for article in corpus_kk_test:
    answer = lng_id_words(article, freqs_words_sort_list, lngs)
    if answer == 'Kazakh':
        kk += 1

Точность классификации казахских текстов

In [381]:
accuracy_kk = kk / len(wiki_texts_test['kk'])
accuracy_kk

1.0

In [382]:
corpus_uk_test = wiki_texts_test['uk']
uk = 0
for article in corpus_uk_test:
    answer = lng_id_words(article, freqs_words_sort_list, lngs)
    if answer == 'Ukranian':
        uk += 1

Точность классификации украинских текстов

In [383]:
accuracy_uk = uk / len(wiki_texts_test['uk'])
accuracy_uk

1.0

In [386]:
corpus_be_test = wiki_texts_test['be']
be = 0
for article in corpus_be_test:
    answer = lng_id_words(article, freqs_words_sort_list, lngs)
    if answer == 'Belorussian':
        be += 1

In [387]:
accuracy_be = be / len(wiki_texts_test['be'])
accuracy_be

0.9789473684210527

In [388]:
corpus_fr_test = wiki_texts_test['fr']
fr = 0
for article in corpus_fr_test:
    answer = lng_id_words(article, freqs_words_sort_list, lngs)
    if answer == 'French':
        fr += 1

In [389]:
accuracy_fr = fr / len(wiki_texts_test['fr'])
accuracy_fr

1.0

Суммарная точность:

In [390]:
accuracy = (kk + uk + be + fr) / len(wiki_texts_test['kk'] + wiki_texts_test['uk'] + wiki_texts_test['be'] + wiki_texts_test['fr'])
accuracy

0.9947368421052631

## Метод 2. N-граммы

Создадим функцию, которая преобразовывает строку text в массив N-грамм заданной длины N.

In [344]:
from itertools import islice, tee

def make_ngrams(text, N=3):
    ngrams = zip(*(islice(seq, index, None) for index, seq in enumerate(tee(text, N))))
    ngrams = [''.join(x) for x in ngrams]
    return ngrams

Теперь создадим частотные словари n-грамм аналогично первому методу.

In [346]:
freqs_grams_kk = collections.defaultdict(lambda: 0)

for article in corpus_kk:
    for ngram in make_ngrams(article.replace('\n', '').lower()):
        freqs_grams_kk[ngram] += 1

for ngram in sorted(freqs_grams_kk, key=lambda n: freqs_grams_kk[n], reverse=True)[:50]:
    print('{}\t{}'.format(freqs_grams_kk[ngram], ngram))

In [335]:
freqs_grams_uk = collections.defaultdict(lambda: 0)

for article in corpus_uk:
    for ngram in make_ngrams(article.replace('\n', '').lower()):
        freqs_grams_uk[ngram] += 1

for ngram in sorted(freqs_grams_uk, key=lambda n: freqs_grams_uk[n], reverse=True)[:50]:
    print('{}\t{}'.format(freqs_grams_uk[ngram], ngram))

In [337]:
freqs_grams_be = collections.defaultdict(lambda: 0)

for article in corpus_be:
    for ngram in make_ngrams(article.replace('\n', '').lower()):
        freqs_grams_be[ngram] += 1

for ngram in sorted(freqs_grams_be, key=lambda n: freqs_grams_be[n], reverse=True)[:50]:
    print('{}\t{}'.format(freqs_grams_be[ngram], ngram))

In [339]:
freqs_grams_fr = collections.defaultdict(lambda: 0)

for article in corpus_fr:
    for ngram in make_ngrams(article.replace('\n', '').lower()):
        freqs_grams_fr[ngram] += 1

for ngram in sorted(freqs_grams_fr, key=lambda n: freqs_grams_fr[n], reverse=True)[:50]:
    print('{}\t{}'.format(freqs_grams_fr[ngram], ngram))

Почистим аналогично первому методу списки от N-грамм, которые встречаются больше, чем в 1 языке. Таким образом, у нас останутся только лингвоспецифические N-граммы.

In [348]:
freqs_grams_list = [freqs_grams_kk, freqs_grams_uk, freqs_grams_be, freqs_grams_fr]

In [349]:
for i in range(0, len(freqs_grams_list)):
    grams = list(freqs_grams_list[i].keys())
    k = 0
    while k != len(grams):
        gram = grams[k]
        for j in range(0, len(freqs_grams_list)):
            if freqs_grams_list[j] != freqs_grams_list[i]:
                if gram in freqs_grams_list[j]:
                    del freqs_grams_list[j][gram]
                    flag = 'T'
        if flag == 'T':
            del freqs_grams_list[i][gram]
            flag = 'F'
        k += 1

Сделаем из них словари по 100 самых частотных N-грамм

In [353]:
freqs_grams_kk_sort = freqs_sort(freqs_grams_kk, 100)
freqs_grams_uk_sort = freqs_sort(freqs_grams_uk, 100)
freqs_grams_be_sort = freqs_sort(freqs_grams_be, 100)
freqs_grams_fr_sort = freqs_sort(freqs_grams_fr, 100)

### Шаг 6. Определить язык

In [355]:
def lng_id_grams(text, freqs, lngs): #freqs - все имеющиеся частотные словари
    count_freqs = [0]*len(freqs)
    text_grammed = make_ngrams(text.replace('\n', '').lower())
    for i in text_grammed:
        for j in range(0, len(freqs)):
            if i in freqs[j]:
                count_freqs[j] += 1
    return lngs[count_freqs.index(max(count_freqs))]

### Шаг 7. Проверить метод на тестовой выборке

In [356]:
freqs_grams_sort_list = [freqs_grams_kk_sort, freqs_grams_uk_sort, freqs_grams_be_sort, freqs_grams_fr_sort]
lngs = ['Kazakh', 'Ukranian', 'Belorussian', 'French']

In [357]:
kk = 0
for article in corpus_kk_test:
    answer = lng_id_grams(article, freqs_grams_sort_list, lngs)
    if answer == 'Kazakh':
        kk += 1

In [358]:
accuracy_kk = kk / len(wiki_texts_test['kk'])
accuracy_kk

1.0

In [359]:
uk = 0
for article in corpus_uk_test:
    answer = lng_id_grams(article, freqs_grams_sort_list, lngs)
    if answer == 'Ukranian':
        uk += 1

In [360]:
accuracy_uk = uk / len(wiki_texts_test['uk'])
accuracy_uk

1.0

In [361]:
be = 0
for article in corpus_be_test:
    answer = lng_id_grams(article, freqs_grams_sort_list, lngs)
    if answer == 'Belorussian':
        be += 1

In [362]:
accuracy_be = be / len(wiki_texts_test['be'])
accuracy_be

1.0

In [363]:
fr = 0
for article in corpus_fr_test:
    answer = lng_id_grams(article, freqs_grams_sort_list, lngs)
    if answer == 'French':
        fr += 1

In [364]:
accuracy_be = be / len(wiki_texts_test['be'])
accuracy_be

1.0

Суммарная точност метода N-грамм:

In [365]:
accuracy = (kk + uk + be + fr) / len(wiki_texts_test['kk'] + wiki_texts_test['uk'] + wiki_texts_test['be'] + wiki_texts_test['fr'])
accuracy

1.0

# Итоги

Программа, определяющая язык текста двумя способами

In [395]:
def lng_id_words_grams(text, lngs, freqs_words=freqs_words_sort_list, freqs_grams=freqs_grams_sort_list):
    count_freqs_words = [0]*len(freqs_words)
    count_freqs_grams = [0]*len(freqs_grams)
    text_grammed = make_ngrams(text.replace('\n', '').lower())
    for i in text_grammed:
        for j in range(0, len(freqs_grams)):
            if i in freqs_grams[j]:
                count_freqs_grams[j] += 1
    text_tokenized = tokenize(text.replace('\n', '').lower())
    for i in text_tokenized:
        for j in range(0, len(freqs_words)):
            if i in freqs_words[j]:
                count_freqs_words[j] += 1
    return lngs[count_freqs_grams.index(max(count_freqs_grams))], lngs[count_freqs_words.index(max(count_freqs_words))]

Метод N-грамм работает лучше. 
На тестовой выборке в 500 документов метод частотных слов дает точность 0.9947368421052631, 
а метод N-грамм - 1.0.

In [396]:
for article in corpus_be_test:
    print(lng_id_words_grams(article, lngs))

('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Belorussian', 'Belorussian')
('Beloru